File: 2019/go/day2/day2.go
package day2

import (
	"fmt"
	"io/ioutil"
	"log"
	"strconv"
	"strings"
)

// Part1 processes intcode and returns first value.
func Part1() {
	nums := runIntcode(readInput("day2/input.txt"), 12, 2)
	fmt.Println(nums[0])
}

// Part2 processes combinations of intcodes and finds a specific first value.
func Part2() {
	fmt.Println("Part 2")
}

func main() {
	Part1()
	Part2()
}

func readInput(file string) []int {
	buf, err := ioutil.ReadFile(file)
	if err != nil {
		log.Fatal(err)
	}

	data := strings.Split(string(buf), ",")
	out := make([]int, len(data))
	for i := range data {
		out[i], _ = strconv.Atoi(data[i])
	}

	return out
}

func runIntcode(nums []int, noun int, verb int) []int {
	nums[1] = noun
	nums[2] = verb

	x := 0
	for x < len(nums) {
		if nums[x] == 99 {
			break
		}

		a := nums[x+1]
		b := nums[x+2]
		c := nums[x+3]

		if nums[x] == 1 {
			nums[c] = nums[a] + nums[b]
		} else if nums[x] == 2 {
			nums[c] = nums[a] * nums[b]
		} else {
			fmt.Printf("Unexpected optcode: %d", nums[x])
		}

		x += 4
	}

	return nums
}

File: 2019/ts/day2/day2.ts
import * as fs from "fs";

const part1 = () => {
  let nums = runIntcode(readInput('day2/input.txt'));
  console.log(nums[0]);
}

const part2 = () => {
  console.log("Part 2");
}

const readInput = (file: string): number[] => {
  return fs.readFileSync(file)
    .toString()
    .trim()
    .split(",")
    .map(Number);
}

const runIntcode = (nums: number[], noun: number = 12, verb: number = 2): number[] => {
  let x = 0;
  nums[1] = noun;
  nums[2] = verb;

  while (x < nums.length) {
    if (nums[x] == 99) {
      break;
    }

    let a = nums[x + 1];
    let b = nums[x + 2];
    let c = nums[x + 3];

    if (nums[x] == 1) {
      nums[c] = nums[a] + nums[b];
    } else if (nums[x] == 2) {
      nums[c] = nums[a] * nums[b];
    } else {
      console.log("Unexpected optcode", nums[x]);
    }
    x += 4;
  }

  return nums;
}

export { part1, part2 };